numba/npyufunc/_internal.c

```diff
@@ -104,6 +105,557 @@ PyTypeObject PyUFuncCleaner_Type = {
+        if (nin_obj) {
+            nin = (int)PyNumber_AsSsize_t(nin_obj, NULL);
+        }
+        Py_XDECREF(nin_obj);
+    } else {
+        nin = (int)_get_nin(py_func_obj);
+        if ((nin < 0) || (PyErr_Occurred())) {
+            Py_XDECREF(py_func_obj);
+            return -1;
+        }
+    }
+    if (PyMapping_HasKeyString(kws, "nout")) {
+        PyObject * nout_obj = PyMapping_GetItemString(kws, "nout");
+        if (nout_obj) {
+            nout = (int)PyNumber_AsSsize_t(nout_obj, NULL);
+        }
+        Py_XDECREF(nout_obj);
+    }
+
+    /* Construct the UFunc. */
+    tmp = PyObject_GetAttrString(py_func_obj, "__name__");
+    if (tmp) {
+#if PY_MAJOR_VERSION >= 3
+        name = PyUnicode_AsUTF8(tmp);
+#else
+        name = PyString_AsString(tmp);
+#endif
+    }
+    Py_XDECREF(tmp);
+    tmp = PyObject_GetAttrString(py_func_obj, "__doc__");
+    if (tmp && (tmp != Py_None)) {
+#if PY_MAJOR_VERSION >= 3
+        doc = PyUnicode_AsUTF8(tmp);
+#else
+        doc = PyString_AsString(tmp);
+#endif
+    }
+    Py_XDECREF(tmp);
+    tmp = NULL;
+    Py_XDECREF(py_func_obj);
+    py_func_obj = NULL;
+    if (!name) {
+        return -1;
+    }
+    ufunc = (PyUFuncObject *)PyUFunc_FromFuncAndData(NULL, NULL, NULL, 0,
+                                                     nin, nout, identity,
+                                                     name, doc, 0);
+    if (!ufunc) {
+        return -1;
+    }
+
+    /* Construct a keepalive list if none was given. */
+    if (!keepalive) {
+        keepalive = PyList_New(0);
+        if (!keepalive) {
+            Py_XDECREF(ufunc);
+            return -1;
+        }
+    } else {
+        Py_INCREF(keepalive);
+    }
+
+    tmp = self->dispatcher;
+    Py_INCREF(dispatcher);
+    self->dispatcher = dispatcher;
+    Py_XDECREF(tmp);
+
+    tmp = (PyObject*)self->ufunc;
+    self->ufunc = ufunc;
+    Py_XDECREF(tmp);
+
+    tmp = self->keepalive;
+    /* Already incref'ed, either by PyList_New(), or else clause, both above. */
+    self->keepalive = keepalive;
+    Py_XDECREF(tmp);
+
+    self->frozen = 0;
+
+    return 0;
+}
+
+static PyMemberDef dufunc_members[] = {
+    {"dispatcher", T_OBJECT_EX, offsetof(PyDUFuncObject, dispatcher), 0,
+         "Dispatcher object for the core Python function."},
+    {"ufunc", T_OBJECT_EX, offsetof(PyDUFuncObject, ufunc), 0,
+         "Numpy Ufunc for the dynamic ufunc."},
+    {"keepalive", T_OBJECT_EX, offsetof(PyDUFuncObject, keepalive), 0,
+         "List of objects to keep alive during life of dufunc."},
+};
+
+/* ____________________________________________________________
+ * Shims to expose ufunc methods.
+ */
+
+static struct _ufunc_dispatch {
+    PyCFunctionWithKeywords ufunc_reduce;
+    PyCFunctionWithKeywords ufunc_accumulate;
+    PyCFunctionWithKeywords ufunc_reduceat;
+    PyCFunctionWithKeywords ufunc_outer;
+    PyCFunction ufunc_at;
+} ufunc_dispatch = {NULL, NULL, NULL, NULL, NULL};
```

Comment:
I'm not going to bother to remove code if has a defensive intent, even if the compiler turns it into a NOP.